pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+
+
+Internally, in Python, ``CurrentUser`` here is equivalent to:
+
+.. code-block::
+
+    CurrentUser = Annotated[str, doc("A user name"), doc("The current user name")]
+
+
+For an implementation that supports the last ``doc()`` appearance, the above
+example would be equivalent to:
+
+.. code-block::
+
+    def hi(
+      to: Annotated[str, doc("The current user name")],
+    ) -> None: ...
+
+
+Early Adopters and Older Python Versions
+========================================
+
+For older versions of Python and early adopters of this proposal, ``doc()`` and
+``DocInfo`` can be imported from the ``typing_extensions`` package.
+
+.. code-block::
+
+    from typing import Annotated
+
+    from typing_extensions import doc
+
+
+    def hi(
+      to: Annotated[str, doc("The current user name")],
+    ) -> None: ...
+
+
+Rejected Ideas
+==============
+
+
+Standardize Current Docstrings
+------------------------------
+
+A possible alternative would be to support and try to push as a standard one of the
+existing docstring formats. But that would only solve the standardization.
+
+It wouldn't solve any of the other problems, like getting editor support
+(syntax checks) for library authors, the distance and duplication of information
+between a parameter definition and its documentation in the docstring, etc.
+
+
+Extra Metadata and Decorator
+----------------------------
+
+An earlier version of this proposal included several parameters to indicate whether
+an object is discouraged from use, what exceptions it may raise, etc.
+To allow also deprecating functions and classes, it was also expected
+that ``doc()`` could be used as a decorator. But this functionality is covered
+by ``typing.deprecated()`` in :pep:`702`, so it was dropped from this proposal.
+
+A way to declare additional information could still be useful in the future,
+but taking early feedback on this document, all that was postponed to future
+proposals.
+
+This also shifts the focus from an all-encompasing function ``doc()``
+with multiple parameters to multiple composable functions, having ``doc()``
+handle one single use case: additional documentation in ``Annotated``.
+
+This design change also allows better interoperability with other proposals
+like ``typing.deprecated()``, as in the future it could be considered to
+allow having ``typing.deprecated()`` also in ``Annotated`` to deprecate
+individual parameters, coexisting with ``doc()``.
+
+
+Open Issues
+===========
+
+
+Verbosity
+---------
+
+The main argument against this would be the increased verbosity.
+
+Nevertheless, this verbosity would not affect end users as they would not see the
+internal code using ``typing.doc()``.
+
+And the cost of dealing with the additional verbosity would only be carried
+by those library maintainers that decide to opt-in into this feature.
+
+Any authors that decide not to adopt it, are free to continue using docstrings
+with any particular format they decide, no docstrings at all, etc.
```

Comment:
See my point above. If this becomes the blessed solution as you hope, library maintainers will face pressure to adopt it -- with the drawbacks of verbosity etc that you mention.